src/internal/kopia/upload.go

```diff
@@ -235,38 +235,229 @@ func (cp *corsoProgress) get(k string) *itemDetails {
 	return cp.pending[k]
 }
 
-func collectionEntries(
-	ctx context.Context,
-	ctr func(context.Context, fs.Entry) error,
-	streamedEnts data.BackupCollection,
-	progress *corsoProgress,
-) (map[string]struct{}, error) {
-	if streamedEnts == nil {
-		return nil, nil
+// These define a small state machine as to which source to return an entry from
+// next. Since these are in-memory only values we can use iota. Phases are
+// traversed in the order defined unless the underlying data source isn't
+// present. If an underlying data source is missing, the non-pre/post phase
+// associated with that data source is skipped.
+//
+// Since some phases require initialization of the underlying data source we
+// insert additional phases to allow that. Once initialization is completed the
+// phase should be updated to the next phase.
+//
+// A similar tactic can be used to handle tearing down resources for underlying
+// data sources if needed.
+const (
+	initPhase = iota
+	staticEntsPhase
+	preStreamEntsPhase
+	streamEntsPhase
+	postStreamEntsPhase
+	preBaseDirEntsPhase
+	baseDirEntsPhase
+	postBaseDirEntsPhase
+	terminationPhase
+)
+
+type corsoDirectoryIterator struct {
+	ctx    context.Context
+	params snapshotParams
+	// staticEnts is the set of fs.StreamingDirectory child directories that we've
+	// generated based on the collections passed in. These entries may or may not
+	// contain an underlying data.BackupCollection (depending on what was passed
+	// in) and may or may not contain an fs.Directory (depending on hierarchy
+	// merging).
+	staticEnts       []fs.Entry
+	globalExcludeSet prefixmatcher.StringSetReader
+	progress         *corsoProgress
+
+	// endSpan is the callback to stop diagnostic span collection for iteration.
+	endSpan func()
+
+	// seenEnts contains the encoded names of entries that we've already streamed
+	// so we can skip returning them again when looking at base entries.
+	seenEnts map[string]struct{}
+	// locationPath contains the human-readable location of the underlying
+	// collection.
+	locationPath *path.Builder
+
+	// excludeSet is the individual exclude set to use for the longest prefix for
+	// this iterator.
```

Comment:
This comment doesn't entirely make sense.  It makes it sound like we use it to exclude the longest prefix.  But that doesn't seem right.